addons/sale_coupon/models/sale_order.py

```diff
@@ -90,14 +90,22 @@ def _get_reward_values_product(self, program):
         # Remove needed quantity from reward quantity if same reward and rule product
         if program._get_valid_products(program.reward_product_id):
             # number of times the program should be applied
-            program_in_order = max_product_qty // (program.rule_min_quantity + program.reward_product_quantity)
-            # multipled by the reward qty
-            reward_product_qty = program.reward_product_quantity * program_in_order
+            program_in_order = (
+                max_product_qty // (program.rule_min_quantity + program.reward_product_quantity)
+            )
             # do not give more free reward than products
-            reward_product_qty = min(reward_product_qty, total_qty)
+            program_in_order = min(program_in_order, (total_qty // program.reward_product_quantity))
+            # multiplied by the reward qty
+            reward_product_qty = program.reward_product_quantity * program_in_order
             if program.rule_minimum_amount:
-                order_total = sum(order_lines.mapped('price_total')) - (program.reward_product_quantity * program.reward_product_id.lst_price)
-                reward_product_qty = min(reward_product_qty, order_total // program.rule_minimum_amount)
+                min_amount_with_reward = (
+                    program.rule_minimum_amount
+                    + program.reward_product_quantity * program.reward_product_id.lst_price
+                )
+                program_in_order = sum(order_lines.mapped('price_total')) // min_amount_with_reward
+                reward_product_qty = min(
+                    reward_product_qty, program_in_order * program.reward_product_quantity
+                )
```

Comment:
Ok, proposition (I just moved some lines, and renamed for clarity’s sake): what would you think of something like this to avoid the repetition?

```py
        if program._get_valid_products(program.reward_product_id):
            # number of times the program could be applied
            nbr_program_by_qtt = (
                max_product_qty // (program.rule_min_quantity + program.reward_product_quantity)
            )
            # Ensure the minimum amount for the reward is met
            min_amount_with_reward = (
                program.rule_minimum_amount
                + program.reward_product_quantity * program.reward_product_id.lst_price
            )
            nbr_program_by_amount = sum(order_lines.mapped('price_total')) // min_amount_with_reward
            # do not give more free reward than products
            nbr_program_less_than_products = total_qty // program.reward_product_quantity
            program_in_order = min(
                nbr_program_by_qtt, nbr_program_by_amount, nbr_program_less_than_products
            )
            # multiplied by the reward qty
            reward_product_qty = program_in_order * program.reward_product_quantity
```

Reply:
great idea, thank you!